### PR TITLE
autodoc: Show full traceback and exception type in warnings

### PR DESCRIPTION
**Pull request by eric-wieser**
Subject: <short purpose of this pull request>

### Feature or Bugfix

- Feature

### Purpose
This should help debug #7384

### CODE REVIEW

File: sphinx/ext/autodoc/__init__.py [hunk 1]
```diff
@@ -751,7 +751,7 @@ def generate(self, more_content: Any = None, real_modname: str = None,
             # be cached anyway)
             self.analyzer.find_attr_docs()
         except PycodeError as err:
-            logger.debug('[autodoc] module analyzer failed: %s', err)
+            logger.debug('[autodoc] module analyzer failed:', excinfo=True)
```
Comment (eric-wieser) [inline sphinx/ext/autodoc/__init__.py]:
```suggestion
            logger.debug('[autodoc] module analyzer failed:', exc_info=True)
```

**Comment by tk0miya** [comment]
+1

File: sphinx/ext/autodoc/__init__.py [hunk 2]
```diff
@@ -397,8 +397,8 @@ def format_signature(self, **kwargs: Any) -> str:
                     # retry without arguments for old documenters
                     args = self.format_args()
             except Exception as err:
```
Comment (eric-wieser) [inline sphinx/ext/autodoc/__init__.py]:
```suggestion
            except Exception:
```

File: sphinx/ext/autodoc/__init__.py [hunk 3]
```diff
@@ -751,7 +751,7 @@ def generate(self, more_content: Any = None, real_modname: str = None,
             # be cached anyway)
             self.analyzer.find_attr_docs()
         except PycodeError as err:
```
Comment (eric-wieser) [inline sphinx/ext/autodoc/__init__.py]:
```suggestion
        except PycodeError:
```

File: sphinx/ext/autodoc/__init__.py [hunk 4]
```diff
@@ -396,9 +396,9 @@ def format_signature(self, **kwargs: Any) -> str:
                 except TypeError:
                     # retry without arguments for old documenters
                     args = self.format_args()
-            except Exception as err:
-                logger.warning(__('error while formatting arguments for %s: %s') %
-                               (self.fullname, err), type='autodoc')
+            except Exception:
+                logger.warning(__('error while formatting arguments for %s:') %
+                               self.fullname, type='autodoc', exc_info=True)
```
Comment (tk0miya) [inline sphinx/ext/autodoc/__init__.py:401]: Note: I thought showing full stacktrace is a bit noisy for users at first. On the other hand, it is very important for developers including extension's authors. In addition, `format_args()` is one of the handlers that is unlikely to throw an exception. So +1 for this change.

**Comment by tk0miya** [comment]
Just merged.

File: sphinx/ext/autodoc/__init__.py (hunk 4 above)
Comment (eric-wieser) [reply sphinx/ext/autodoc/__init__.py:401]: It might be reasonable to respect a verbose flag here, as some future work.

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:401]: Absolutely!

Comment (tk0miya) [reply sphinx/ext/autodoc/__init__.py:401]: I made a patch for this purpose #7608

**Pull request merged**